diff --git a/linux-x86/1.82.0/lib/rustlib/src/rust/library/core/src/fmt/mod.rs b/linux-x86/1.82.0/lib/rustlib/src/rust/library/core/src/fmt/mod.rs
index 45c2b6a6a0f73efa7436d50855c3dd884a912794..4d0b9bce215e27eec430148a02a8236bc2a0d6d2 100644
--- a/linux-x86/1.82.0/lib/rustlib/src/rust/library/core/src/fmt/mod.rs
+++ b/linux-x86/1.82.0/lib/rustlib/src/rust/library/core/src/fmt/mod.rs
@@ -7,7 +7,7 @@ use crate::char::EscapeDebugExtArgs;
 use crate::marker::PhantomData;
 use crate::num::fmt as numfmt;
 use crate::ops::Deref;
-use crate::{iter, mem, result, str};
+use crate::{iter, result, str};
 
 mod builders;
 #[cfg(not(no_fp_fmt_parse))]
diff --git a/linux-x86/1.82.0/lib/rustlib/src/rust/library/core/src/fmt/rt.rs b/linux-x86/1.82.0/lib/rustlib/src/rust/library/core/src/fmt/rt.rs
index f29ac99b292556240ed4f32883460301ae06cfed..30e1db2022661401544ae7f7239387bd8add9ef2 100644
--- a/linux-x86/1.82.0/lib/rustlib/src/rust/library/core/src/fmt/rt.rs
+++ b/linux-x86/1.82.0/lib/rustlib/src/rust/library/core/src/fmt/rt.rs
@@ -93,61 +93,92 @@ pub struct Argument<'a> {
     ty: ArgumentType<'a>,
 }
 
-#[rustc_diagnostic_item = "ArgumentMethods"]
-impl<'a> Argument<'a> {
-    #[inline(always)]
-    fn new<'b, T>(x: &'b T, f: fn(&T, &mut Formatter<'_>) -> Result) -> Argument<'b> {
+macro_rules! argument_new {
+    ($t:ty, $x:expr, $f:expr) => {
         Argument {
             // INVARIANT: this creates an `ArgumentType<'b>` from a `&'b T` and
             // a `fn(&T, ...)`, so the invariant is maintained.
             ty: ArgumentType::Placeholder {
-                value: NonNull::from(x).cast(),
-                // SAFETY: function pointers always have the same layout.
-                formatter: unsafe { mem::transmute(f) },
+                value: NonNull::<$t>::from($x).cast(),
+                // The Rust ABI considers all pointers to be equivalent, so transmuting a fn(&T) to
+                // fn(NonNull<()>) and calling it with a NonNull<()> that points at a T is allowed.
+                // However, the CFI sanitizer does not allow this, and triggers a crash when it
+                // happens.
+                //
+                // To avoid this crash, we use a helper function when CFI is enabled. To avoid the
+                // cost of this helper function (mainly code-size) when it is not needed, we
+                // transmute the function pointer otherwise.
+                //
+                // This is similar to what the Rust compiler does internally with vtables when KCFI
+                // is enabled, where it generates trampoline functions that only serve to adjust the
+                // expected type of the argument. `ArgumentType::Placeholder` is a bit like a
+                // manually constructed trait object, so it is not surprising that the same approach
+                // has to be applied here as well.
+                //
+                // It is still considered problematic (from the Rust side) that CFI rejects entirely
+                // legal Rust programs, so we do not consider anything done here a stable guarantee,
+                // but meanwhile we carry this work-around to keep Rust compatible with CFI and
+                // KCFI.
+                #[cfg(not(any(sanitize = "cfi", sanitize = "kcfi")))]
+                formatter: {
+                    let f: fn(&$t, &mut Formatter<'_>) -> Result = $f;
+                    // SAFETY: This is only called with `value`, which has the right type.
+                    unsafe { core::mem::transmute(f) }
+                },
+                #[cfg(any(sanitize = "cfi", sanitize = "kcfi"))]
+                formatter: |ptr: NonNull<()>, fmt: &mut Formatter<'_>| {
+                    let func = $f;
+                    // SAFETY: This is the same type as the `value` field.
+                    let r = unsafe { ptr.cast::<$t>().as_ref() };
+                    (func)(r, fmt)
+                },
                 _lifetime: PhantomData,
             },
         }
-    }
+    };
+}
 
+#[rustc_diagnostic_item = "ArgumentMethods"]
+impl<'a> Argument<'a> {
     #[inline(always)]
     pub fn new_display<'b, T: Display>(x: &'b T) -> Argument<'_> {
-        Self::new(x, Display::fmt)
+        argument_new!(T, x, <T as Display>::fmt)
     }
     #[inline(always)]
     pub fn new_debug<'b, T: Debug>(x: &'b T) -> Argument<'_> {
-        Self::new(x, Debug::fmt)
+        argument_new!(T, x, <T as Debug>::fmt)
     }
     #[inline(always)]
     pub fn new_debug_noop<'b, T: Debug>(x: &'b T) -> Argument<'_> {
-        Self::new(x, |_, _| Ok(()))
+        argument_new!(T, x, |_, _| Ok(()))
     }
     #[inline(always)]
     pub fn new_octal<'b, T: Octal>(x: &'b T) -> Argument<'_> {
-        Self::new(x, Octal::fmt)
+        argument_new!(T, x, <T as Octal>::fmt)
     }
     #[inline(always)]
     pub fn new_lower_hex<'b, T: LowerHex>(x: &'b T) -> Argument<'_> {
-        Self::new(x, LowerHex::fmt)
+        argument_new!(T, x, <T as LowerHex>::fmt)
     }
     #[inline(always)]
     pub fn new_upper_hex<'b, T: UpperHex>(x: &'b T) -> Argument<'_> {
-        Self::new(x, UpperHex::fmt)
+        argument_new!(T, x, <T as UpperHex>::fmt)
     }
     #[inline(always)]
     pub fn new_pointer<'b, T: Pointer>(x: &'b T) -> Argument<'_> {
-        Self::new(x, Pointer::fmt)
+        argument_new!(T, x, <T as Pointer>::fmt)
     }
     #[inline(always)]
     pub fn new_binary<'b, T: Binary>(x: &'b T) -> Argument<'_> {
-        Self::new(x, Binary::fmt)
+        argument_new!(T, x, <T as Binary>::fmt)
     }
     #[inline(always)]
     pub fn new_lower_exp<'b, T: LowerExp>(x: &'b T) -> Argument<'_> {
-        Self::new(x, LowerExp::fmt)
+        argument_new!(T, x, <T as LowerExp>::fmt)
     }
     #[inline(always)]
     pub fn new_upper_exp<'b, T: UpperExp>(x: &'b T) -> Argument<'_> {
-        Self::new(x, UpperExp::fmt)
+        argument_new!(T, x, <T as UpperExp>::fmt)
     }
     #[inline(always)]
     pub fn from_usize(x: &usize) -> Argument<'_> {
@@ -159,11 +190,6 @@ impl<'a> Argument<'a> {
     /// # Safety
     ///
     /// This argument must actually be a placeholder argument.
-    ///
-    // FIXME: Transmuting formatter in new and indirectly branching to/calling
-    // it here is an explicit CFI violation.
-    #[allow(inline_no_sanitize)]
-    #[no_sanitize(cfi, kcfi)]
     #[inline(always)]
     pub(super) unsafe fn fmt(&self, f: &mut Formatter<'_>) -> Result {
         match self.ty {
diff --git a/linux-x86/1.82.0/lib/rustlib/src/rust/library/core/src/lib.rs b/linux-x86/1.82.0/lib/rustlib/src/rust/library/core/src/lib.rs
index 06a745b690a4b72854ab9f0f8acc7c16892f8f1e..6eb18d4747b02c6036dafbb3b6bdd4f1a493ec70 100644
--- a/linux-x86/1.82.0/lib/rustlib/src/rust/library/core/src/lib.rs
+++ b/linux-x86/1.82.0/lib/rustlib/src/rust/library/core/src/lib.rs
@@ -233,7 +233,6 @@
 #![feature(negative_impls)]
 #![feature(never_type)]
 #![feature(no_core)]
-#![feature(no_sanitize)]
 #![feature(prelude_import)]
 #![feature(repr_simd)]
 #![feature(rustc_allow_const_fn_unstable)]
